src/sinks/splunk_hec/common/acknowledgements.rs

```diff
@@ -0,0 +1,294 @@
+        for ack_id in ack_ids {
+            if let Some((_, ack_event_status_sender)) = self.acks.remove(ack_id) {
+                let _ = ack_event_status_sender.send(EventStatus::Delivered);
+                debug!("Finalized ack id {:?}", ack_id);
+            }
+        }
+    }
+
+    /// Builds an ack query body with stored ack ids
+    fn get_ack_query_body(&mut self) -> HecAckStatusRequest {
+        self.clear_expired_ack_ids();
+        HecAckStatusRequest {
+            acks: self.acks.keys().copied().collect::<Vec<u64>>(),
+        }
+    }
+
+    /// Decrements retry count on all stored ack ids by 1
+    fn decrement_retries(&mut self) {
+        for (retries, _) in self.acks.values_mut() {
+            *retries = retries.checked_sub(1).unwrap_or(0);
+        }
+    }
+
+    /// Removes all expired ack ids (those with a retry count of 0).
+    fn clear_expired_ack_ids(&mut self) {
+        self.acks.retain(|_, (retries, _)| *retries > 0);
+    }
+
+    // Sends an ack status query request to Splunk HEC
+    async fn send_ack_query_request(
+        &mut self,
+        request_body: &HecAckStatusRequest,
+    ) -> Result<HecAckStatusResponse, HecAckApiError> {
+        self.decrement_retries();
+        let request_body_bytes =
+            serde_json::to_vec(request_body).map_err(|_| HecAckApiError::ClientBuildRequest)?;
+        let request = self
+            .http_request_builder
+            .build_request(request_body_bytes, "/services/collector/ack")
+            .map_err(|_| HecAckApiError::ClientBuildRequest)?;
+
+        let response = self
+            .client
+            .send(request.map(Body::from))
+            .await
+            .map_err(|_| HecAckApiError::ServerSendQuery)?;
+
+        let status = response.status();
+        if status.is_success() {
+            let response_body = hyper::body::to_bytes(response.into_body())
+                .await
+                .map_err(|_| HecAckApiError::ClientParseResponse)?;
+            serde_json::from_slice::<HecAckStatusResponse>(&response_body)
+                .map_err(|_| HecAckApiError::ClientParseResponse)
+        } else if status.is_client_error() {
+            Err(HecAckApiError::ClientSendQuery)
+        } else {
+            Err(HecAckApiError::ServerSendQuery)
+        }
+    }
+}
+
+pub async fn run_acknowledgements(
+    mut receiver: UnboundedReceiver<(u64, Sender<EventStatus>)>,
+    client: HttpClient,
+    http_request_builder: Arc<HttpRequestBuilder>,
+    indexer_acknowledgements: HecClientAcknowledgementsConfig,
+) {
+    let mut interval = tokio::time::interval(Duration::from_secs(
+        indexer_acknowledgements.query_interval.get() as u64,
+    ));
+    let mut ack_client = HecAckClient::new(
+        indexer_acknowledgements.retry_limit.get(),
+        client,
+        http_request_builder,
+    );
+
+    loop {
+        tokio::select! {
+            _ = interval.tick() => {
+                ack_client.run().await;
```

Comment:
Could this ack client run in parallel with this receive loop to avoid the need for a select, and also avoid issues with delays fetching the acks from delaying receiving new data from the channel?

Reply:
If they ran in parallel, I think we'd need a lock around `HecAckClient.acks`. Running the ack client may require modifying `acks` and receiving new `ack_id`'s also modifies `acks`. Would the required locking offset the benefits of parallelizing?

Reply:
Yes, we would either need a lock or similar. My concern was if there was a delay in the process of fetching the acks, the new ids could pile up waiting behind that process. This doesn't have to be blocking if others think this is a minor concern.

Reply:
Once merged, I'll add a soak test with `splunk_hec source` -> possible transform -> `splunk_hec sink` (the lading splunk generator/blackhole have acking behavior). I'm hoping that will show if there are significant issues.